### PR TITLE
Add Feature Implementation

### LINKED ISSUE #80: Completing Add feature

### PR DESCRIPTION
**Pull request by RyanWork**
Hello, following #78 I think I got the add feature to work well.

Here are the changes:
- In Commit: 26ace275f0320f0239749a34b9d762e65179c151 I added a RequestWrapper model (I seriously couldn't figure out a better name). The RequestWrapper is just a single object that holds an instance of a Course, LearningOutcome, Category, and Program when the user is adding a new entry. This is so we can serialize the data and pass it as a single request body via AJAX (I wasn't sure how to post multiple objects to a single POST request).

- In Commit: c82f9f0f52bef8f4fff35719868599b77dd9c2d0 I moved the table in admin to a fragment so I can dynamically update the table as new requests are made.

- In Commit: 156a4169b524e14f382654206aff88db92d85b1d Just fixed some merge conflicts with Kshamina's Select Year

Closes #80.

### CODE REVIEW

**Review by SupriyaGadigone — APPROVED** [review]
Looks good to me! I just had a question though, I added a course with empty fields and it works, I guess that is the expected functionality? Or should we make an issue to have some errors or message show up that a course needs to be linked to a Category or something?

**Review by K-shamG — COMMENTED** [review]
Is the intention that your creating everything from scratch (course, program and category, etc?) at once? I think the form values should have the option to choose from existing learning outcomes, categories, programs etc or create a new one. (https://select2.org/tagging) 

This also looks like it's just for creating new courses. Is there a reason for this? If it's for creating a new anything, I'm not getting that sense from the form. :O 

And a course can belong to many programs or have many LO's - it doesn't look like I am able to do this with the form given. Please explain what I can and can't do with this feature because it is a little unclear.

**Comment by RyanWork** [comment]
@SupriyaGadigone 
> Looks good to me! I just had a question though, I added a course with empty fields and it works, I guess that is the expected functionality? Or should we make an issue to have some errors or message show up that a course needs to be linked to a Category or something?

The input would be covered through model validation. In terms of adding, I think it works the way it is intended though. We will have extra parameters to check this kind of stuff.

@K-shamG 
>Is the intention that your creating everything from scratch (course, program and category, etc?) at once? I think the form values should have the option to choose from existing learning outcomes, categories, programs etc or create a new one. (https://select2.org/tagging)

I believe Adam mentioned that he wanted to do something like this. During our last milestone I recall that he told the professor we were going to implement some kind of drop down the looks at some of the available options to choose from.

> This also looks like it's just for creating new courses. Is there a reason for this? If it's for creating a new anything, I'm not getting that sense from the form. :O

I think the idea is that the admin should be able to create new anythings? We can't exactly connect our third-party system to Carleton to see the available list of Courses/Programs they have. Maybe we should have a separate integration for creating these kinds of fields? As I said above, this should probably not be intended for creating new things, but building new relationships. It would help if we had dropdowns or tagging.

> And a course can belong to many programs or have many LO's - it doesn't look like I am able to do this with the form given. Please explain what I can and can't do with this feature because it is a little unclear.

No implementation yet? Thonking. Tagging fixes this and we pass a list in the JSON object instead.

I guess there's a lot more work to be done? I was just integrating some of Adam's stuff. I'll go back and revise.

**Comment by adamstaples** [comment]
> Looks good to me! I just had a question though, I added a course with empty fields and it works, I guess that is the expected functionality? Or should we make an issue to have some errors or message show up that a course needs to be linked to a Category or something?

As what Ryan mentioned, I think implementing more parameter checking, is ideal, I 'll get on that!

>Is the intention that your creating everything from scratch (course, program and category, etc?) at once? I think the form values should have the option to choose from existing learning outcomes, categories, programs etc or create a new one. (https://select2.org/tagging)

Following on from what Ryan mentioned. The idea is that you can pick from any existing courses programs etc. but with the ability to also create your own. In terms of UI I'm thinking just a little + button next to each field might work out. Modal inside a Modal might not be ideal but probably best option?

> And a course can belong to many programs or have many LO's - it doesn't look like I am able to do this with the form given. Please explain what I can and can't do with this feature because it is a little unclear

This was the part that I was most worried about. In my head if the modal window to add has drop down menus that show all existing courses, program, etc. If we we select let's say a course, maybe some of those menus gray out to imply they aren't possible to be added to a course?

 For example: I select a course from the drop down menu, year and course code grey out, because I can't add another one of those options to a course, but LO's and programs stay active and then when one is selected and the add button is pressed those are added to the course picked.

I dunno if that logic was easy to follow it's pretty hard to explain.

**Comment by K-shamG** [comment]
> This was the part that I was most worried about. In my head if the modal window to add has drop down menus that show all existing courses, program, etc. If we we select let's say a course, maybe some of those menus gray out to imply they aren't possible to be added to a course?
> 
> For example: I select a course from the drop down menu, year and course code grey out, because I can't add another one of those options to a course, but LO's and programs stay active and then when one is selected and the add button is pressed those are added to the course picked.
> 
> I dunno if that logic was easy to follow it's pretty hard to explain.

Why not just have different add buttons for each type of thing? I think that's the easiest way. You could even have a Add dropdown that indicates what type of thing you want to add (course, LO, program etc) and then a different modal for each.

**Comment by adamstaples** [comment]
In commit https://github.com/sysc4806/SYSC4806-WDWH/commit/8030603274ba60031f1b72d7708a4b741404d999:

Made the categories, learning O and program fields select similar to the user page,  and also made the add button not pressable unless everything is filled in correctly.
Buttons for new categories, learning Os and programs added, but functionality to be added tonight!

**Comment by K-shamG** [comment]
I'm wondering - why did you choose to have just one Add button and then more add buttons in the modal. Why not just have four separate add buttons for each type where your one Add button is. 

Aka in this header, have 4 add buttons instead of one. This will remove the complexity of modals within modals.
image

Or you could still have one button and use dropdowns: 
image
https://getbootstrap.com/docs/4.3/components/dropdowns/ 

Also when I added a test course and went to add another, the dropdown had duplicates for some reason: 
image

Another problem is if I add a category, I am still able to add an LO that doesn't belong to that category. E.g. I can add programming and then add the LO "Design Software System", but it's actual category isn't there.

**Comment by K-shamG** [comment]
@adamstaples Just to give you an idea of what I was thinking for the add feature. (@RyanWork does this make sense to you?)

There should be one dropdown add button with the options of : course, program, category, LO

1. Course
- Name
- Year 
- Code
- List of LOs (multiple select)
- Program 
^ LO and program will all choose from existing values. There is no category because Course doesn't have a direct relationship with Category. It is associated to it through the LO's it contains. 

2. Program 
- Name
- List of Courses
^Again, choose from existing courses

3. LO
- Name
- Category

4. Category
- Name
- List of existing LOs

I know it's not the nicest to have to create a new entity separately, but I think this is the simplest way to do it. You could probably combine New Category and LO.

Optional instead of 3 & 4: 
- Category Name
- Multi-select of LOs (can add new as well)

**Comment by adamstaples** [comment]
> Aka in this header, have 4 add buttons instead of one. This will remove the complexity of modals witJust to give you an idea of what I was thinking for the add feature.

This is a great Idea, For now I might just implement 4 buttons, if it looks real bad we can make the drop down menue, but I totally get what you mean now!

> Also when I added a test course and went to add another, the dropdown had duplicates for some reason:

> Another problem is if I add a category, I am still able to add an LO that doesn't belong to that category. E.g. I can add programming and then add the LO "Design Software System", but it's actual category isn't there.

Both of these I think I know how to fix, thanks for bringing them up!

**Comment by adamstaples** [comment]
Commit: https://github.com/sysc4806/SYSC4806-WDWH/commit/b310e8df88df5fd622a07da564344a8aceacec50

Hey Friends, all add features implemented, feel free to go in a check it out. Gonna go eat and go home and then clean up the code and add comments etc. as right now it's a little sad.

**Review by K-shamG — CHANGES_REQUESTED** [review]
Let's say i'm adding a new program with no courses created for it yet. I can't do that because it has to have courses added. 
e.g. if i create "testProgram" and I want to add a new course "testCourse" to it, I can't because neither program nor course allow for that field to be empty. I think one or the other has to allow empty in case of a totally new program

File: src/main/resources/static/js/index.js [hunk 1]
```diff
@@ -90,3 +90,123 @@ function getFilterResults() {
...
+function post(url, data){
+    $.ajax({
+        type: "POST",
+        url: url,
+        data: JSON.stringify(data),
+        dataType : "text",
+        contentType: "application/json",
+        success: function(data) {
+            // Update the table
+            //taken out to reload the page such that new courses are seen.
```
Comment (K-shamG) [inline src/main/resources/static/js/index.js]: Remove commented out stuff if uneeded.

Comment (adamstaples) [reply src/main/resources/static/js/index.js]: Will do, there are a few things like that going to push a commit that cleans up the code a bunch soon

**Comment by adamstaples** [comment]
> e.g. if i create "testProgram" and I want to add a new course "testCourse" to it, I can't because neither program nor course allow for that field to be empty. I think one or the other has to allow empty in case of a totally new program

Very good point to make things easier I'll allow the program to be created with an empty course field!

**Comment by K-shamG** [comment]
> > e.g. if i create "testProgram" and I want to add a new course "testCourse" to it, I can't because neither program nor course allow for that field to be empty. I think one or the other has to allow empty in case of a totally new program
> 
> Very good point to make things easier I'll allow the program to be created with an empty course field!

Perf!! It seems to be working fine other than that :D

**Comment by adamstaples** [comment]
commit: https://github.com/sysc4806/SYSC4806-WDWH/commit/ae545e0dda50d7e0d4131cd223cc22051ce91e41

Just cleaned up the code, think it's good to merge once reviews are done.

**Review by adamstaples — APPROVED** [review]
I think we are all good to merge this!

**Review by K-shamG — APPROVED** [review]
